clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -2730,6 +2731,7 @@ private void handleUpdateRequest(
         // either. So, there is no need to tell the follower replica to do anything.
       }
     } else {
+      blockOnDrainerCapacity();
```

Comment:
Calling`blockOnDrainerCapacity` on almost every record is will impact performance negatively as it calls AtomicLong::get. It will be interesting to see the ingestion perf after this change.

Reply:
Nevermind, it's an array not a map, so getting the aggregate is probably the best way.

Reply:
Maybe feasible to add a function to the `StoreBufferService` along the lines of `boolean isDrainerNearCapacity(ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecord, int subPartition)` which would then be able to call `SBS::getDrainerIndexForConsumerRecord` and look up that specific drainer's capacity. Could also be achieved by passing just a `String consumedTopic` rather than a full record, given a minor refactoring within `SBS`.

Reply:
> We're already examining and updating the capacity for every submission to the drainer, 
IIUC, this happens in the drainer threads which are limited in number; however, we have far more shared consumer threads than the drainer threads. Hence the impact of updating atomic variables millions of times per second might be even more pronounced.